libs/pilight/protocols/433.92/tfa2017.c

```diff
@@ -0,0 +1,246 @@
+				(memcmp(&binary[start[0]], &binary[start[2]], MESSAGE_LENGTH) == 0 ||
+					memcmp(&binary[start[1]], &binary[start[2]], MESSAGE_LENGTH) == 0)) {
+		m=start[2];
+	} else {
+		return;
+	}
+
+	// decode manchester
+	prev = 1;
+	for(x=0;x<MESSAGE_LENGTH;x++) {
+		if (binary[x+m] == 0) {
+			prev = !prev;
+		}
+		msg[x] = prev;
+	}
+	// According to http://www.osengr.org/WxShield/Downloads/Weather-Sensor-RF-Protocols.pdf
+	// the first byte is a fixed id, the second is a rolling code which changes on
+	// battery replacement (both are not used here).
+	// Of the next four bits the first is unused, the next three encode the channel.
+	channel=binToDecRev(msg, 17, 19)+1;
+	// The next twelve bits encode the temperature T
+	// in tenth of degree Fahrenheit with an offset of 40.
+	// The following is a simplification of F=T/10-40 and C=(F-32)*5/9.
+	temperature=(double)binToDecRev(msg, 20, 31)/18.-40.;
+	// The next byte has the relative humidity in percent.
+	humidity=(double)binToDecRev(msg, 32, 39);
+	// The last byte contains a checksum which is not used here.
+
+	struct settings_t *tmp = settings;
+	while(tmp) {
+		if(fabs(tmp->id-channel) < EPSILON){
+			temperature += tmp->temp;
+			humidity += tmp->humi;
+			break;
+		}
+		tmp = tmp->next;
+	}
+
+	snprintf((*message), 255,
+		"{\"id\":%d,\"temperature\":%.1f,\"humidity\":%.1f}",
+		channel, temperature, humidity
+	);
+}
+
+static int checkValues(struct JsonNode *jvalues) {
+	struct JsonNode *jid = NULL;
+
+	if((jid = json_find_member(jvalues, "id"))) {
+		struct settings_t *snode = NULL;
+		struct JsonNode *jchild = NULL;
+		struct JsonNode *jchild1 = NULL;
+		double id = -1;
+		int match = 0;
+
+		jchild = json_first_child(jid);
+		while(jchild) {
+			jchild1 = json_first_child(jchild);
+			while(jchild1) {
+				if(strcmp(jchild1->key, "id") == 0) {
+					id = jchild1->number_;
+				}
+				jchild1 = jchild1->next;
+			}
+			jchild = jchild->next;
+		}
+
+		struct settings_t *tmp = settings;
+		while(tmp) {
+			if(fabs(tmp->id-id) < EPSILON) {
+				match = 1;
+				break;
+			}
+			tmp = tmp->next;
+		}
+
+		if(match == 0) {
+			if((snode = MALLOC(sizeof(struct settings_t))) == NULL) {
+				fprintf(stderr, "out of memory\n");
+				exit(EXIT_FAILURE);
+			}
+			snode->id = id;
+			snode->temp = 0;
+			snode->humi = 0;
+
+			json_find_number(jvalues, "temperature-offset", &snode->temp);
+			json_find_number(jvalues, "humidity-offset", &snode->humi);
+
+			snode->next = settings;
+			settings = snode;
+		}
+	}
+	return 0;
+}
+
+static void gc(void) {
+	struct settings_t *tmp = NULL;
+	while(settings) {
+		tmp = settings;
+		settings = settings->next;
+		FREE(tmp);
+	}
+	if(settings != NULL) {
+		FREE(settings);
+	}
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void tfa2017Init(void) {
+	protocol_register(&tfa2017);
+	protocol_set_id(tfa2017, "tfa2017");
+	protocol_device_add(tfa2017, "tfa2017", "TFA 30.X Temp Hum Sensor Protocol 2017");
```

Comment:
Is the addition of protocol `Protocol 2017` necessary?

Reply:
I just wanted to give it a description different from the string which is in tfa30. Suggestions are welcome.

Reply:
Maybe `tfa3.0-2`?
And `TFA 30.X Temp Hum Sensor Protocol v2`?

Reply:
I am not sure, I think 3.0 is misleading because as far as I understood the 30 refers to the first two digits of the product number (Kat.-Nr.?) as printed on the device which is 30.3208.02 for my sender and 30.3054.10 on the receiver. I will ask in the forum threads maybe there is a good idea for a name and description.